DOC: Add time complexities of csgraph algorithms

czgdp1807 opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->

https://github.com/scipy/scipy/issues/5346

#### What does this implement/fix?
<!--Please explain your changes.-->

- [x] [connected_components](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.connected_components.html#scipy.sparse.csgraph.connected_components)
- [x] [laplacian](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.laplacian.html#scipy.sparse.csgraph.laplacian)
- [x] [shortest_path](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.shortest_path.html#scipy.sparse.csgraph.shortest_path)
- [x] [dijkstra](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.dijkstra.html#scipy.sparse.csgraph.dijkstra)
- [x] [floyd_warshall](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.floyd_warshall.html#scipy.sparse.csgraph.floyd_warshall)
- [x] [bellman_ford](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.bellman_ford.html#scipy.sparse.csgraph.bellman_ford)
- [x] [johnson](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.johnson.html#scipy.sparse.csgraph.johnson)
- [x] [yen](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.yen.html#scipy.sparse.csgraph.yen)
- [x] [breadth_first_order](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.breadth_first_order.html#scipy.sparse.csgraph.breadth_first_order)
- [x] [depth_first_order](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.depth_first_order.html#scipy.sparse.csgraph.depth_first_order)
- [x] [breadth_first_tree](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.breadth_first_tree.html#scipy.sparse.csgraph.breadth_first_tree)
- [x] [depth_first_tree](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.depth_first_tree.html#scipy.sparse.csgraph.depth_first_tree)
- [x] [minimum_spanning_tree](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.minimum_spanning_tree.html#scipy.sparse.csgraph.minimum_spanning_tree)
- [x] [reverse_cuthill_mckee](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.reverse_cuthill_mckee.html#scipy.sparse.csgraph.reverse_cuthill_mckee)
- [x] [maximum_flow](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.maximum_flow.html#scipy.sparse.csgraph.maximum_flow)
- [x] [maximum_bipartite_matching](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.maximum_bipartite_matching.html#scipy.sparse.csgraph.maximum_bipartite_matching)
- [x] [min_weight_full_bipartite_matching](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.min_weight_full_bipartite_matching.html#scipy.sparse.csgraph.min_weight_full_bipartite_matching)
- [x] [structural_rank](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.structural_rank.html#scipy.sparse.csgraph.structural_rank)
- [x] [construct_dist_matrix](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.construct_dist_matrix.html#scipy.sparse.csgraph.construct_dist_matrix)
- [x] ~[csgraph_from_dense](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.csgraph_from_dense.html#scipy.sparse.csgraph.csgraph_from_dense) - Do we need to mention time complexity for this?~
- [x] ~[csgraph_from_masked](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.csgraph_from_masked.html#scipy.sparse.csgraph.csgraph_from_masked) - Do we need to mention time complexity for this?~
- [x] ~[csgraph_masked_from_dense](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.csgraph_masked_from_dense.html#scipy.sparse.csgraph.csgraph_masked_from_dense) - Do we need to mention time complexity for this?~
- [x] ~[csgraph_to_dense](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.csgraph_to_dense.html#scipy.sparse.csgraph.csgraph_to_dense) - Do we need to mention time complexity for this?~
- [x] ~[csgraph_to_masked](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.csgraph_to_masked.html#scipy.sparse.csgraph.csgraph_to_masked) - Do we need to mention time complexity for this?~
- [x] [reconstruct_path](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.reconstruct_path.html#scipy.sparse.csgraph.reconstruct_path)

#### Additional information
<!--Any additional information you think is important.-->

czgdp1807 committed:
DOC: Added time complexities for csgraph algorithms

czgdp1807 committed:
MISC: Styling changes

czgdp1807 commented:
If its hard to review let me know, I will make smaller PRs.

j-bowhay commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -112,6 +112,9 @@ def laplacian(
     symmetric if the input csgraph was symmetric.
 
+    The algorithmic complexity is for a graph with V
+    vertices is O(V^2).
```
Thanks for tackling this, I wonder if these should be monospace
```suggestion
    vertices is ``O(V^2)``.
```

dschult commented on scipy/sparse/csgraph/_reordering.pyx:
```diff
@@ -197,12 +197,17 @@ def structural_rank(graph):
+
+    Notes
+    -----
+    Algorithmic complexity is :math:`O(\lvert E \rvert \sqrt{\lvert V \rvert})`,
```
I don't think you need the `\lvert` here if V is `the number of vertices`. That would be for V being the set of vertices.  BTW: I'm glad to see that you are stating what E and V are.

czgdp1807 replied:
Ah! yes correct. Thanks for pointing out. My bad.

dschult commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -1408,6 +1422,9 @@ def yen(
     Python version.
 
+    The algorithmic complexity for a graph with V number of vertices and E
+    number of edges is O(K*V*(E + V*logV)).
```
What is K here? Is that the arbitrary constant in the big-O notation?

czgdp1807 replied:
`K` is the input argument of `yen`'s algorithm. `yen`'s algorithm finds the K-shortest paths.

dschult reviewed (commented):
Thanks for this...  
Are you getting these complexities based on published papers?
Or based on the actual code used?

czgdp1807 committed:
DOC: Address reviews

czgdp1807 commented:
I addressed the reviews. Actually I am trying to build docs on my machine but I am getting the following error repeatedly (with python 3.11 and python 3.12),

```zsh
reading sources... [100%] tutorial/stats/continuous_t .. tutorial/thread_safety
/Users/czgdp1807/Quansight/scipy/build-install/lib/python3.11/site-packages/scipy/stats/__init__.py:docstring of scipy.stats:47: WARNING: autosummary: stub file not found 'scipy.stats.uniform'. Check your autosummary_generate setting.
/Users/czgdp1807/Quansight/scipy/build-install/lib/python3.11/site-packages/scipy/signal/_ltisys.py:docstring of scipy.signal._ltisys.place_poles:127: WARNING: Exception occurred in plotting scipy-signal-place_poles-1
 from /Users/czgdp1807/Quansight/scipy/doc/source/reference/generated/scipy.signal.place_poles.rst:
Traceback (most recent call last):
  File "/Users/czgdp1807/mambaforge/envs/scipy-dev-py311/lib/python3.11/site-packages/matplotlib/sphinxext/plot_directive.py", line 552, in _run_code
    exec(code, ns)
  File "<string>", line 46, in <module>
  File "/Users/czgdp1807/mambaforge/envs/scipy-dev-py311/lib/python3.11/site-packages/numpy/linalg/_linalg.py", line 2383, in det
    r = _umath_linalg.det(a, signature=signature)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeWarning: divide by zero encountered in det
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
copying assets... 
copying static files... done
copying extra files... done
copying assets: done
writing output... [  9%] reference/generated/scipy.integrate.DOP853.C .. reference/generated/scipy.interpolate.KroghInterpolator
writing output... [ 14%] reference/generated/scipy.interpolate.KroghInterpolator.derivatives .. reference/generated/scipy.linalg
/Users/czgdp1807/Quansight/scipy/build-install/lib/python3.11/site-packages/scipy/stats/__init__.py:docstring of scipy.stats:479:<autosummary>:1: WARNING: py:obj reference target not found: scipy.stats.Uniform
/Users/czgdp1807/Quansight/scipy/doc/source/release/1.15.0-notes.rst:260: WARNING: py:obj reference target not found: scipy.stats.Uniform
/Users/czgdp1807/Quansight/scipy/doc/source/release/1.15.0-notes.rst:266: WARNING: py:obj reference target not found: scipy.stats.Uniform
writing output... [ 68%] reference/generated/scipy.sparse.lil_matrix.real .. reference/generated/scipy.spatial.distance.sokalsne
writing output... [100%] tutorial/stats/continuous_halfcauchy .. tutorial/thread_safety
generating indices... genindex done
writing additional pages... search done
copying images... [  9%] ../build/plot_directive/reference/generated/scipy-interpolate-SmoothSphereBivariateSpline-__call__-1.pn
copying images... [100%] ../build/plot_directive/tutorial/stats/sampling_srou-1.png
dumping search index in English (code: en)... done
dumping object inventory... done
[jupyterlite-sphinx] Running JupyterLite build
[jupyterlite-sphinx] JupyterLite build done
build finished with problems, 5 warnings.
make: *** [html-build] Error 1
``` 

Here's my conda packages information,

<details>

<summary>Conda Packages</summary>

```zsh
(scipy-dev-py311) 19:24:24:~/Quansight/scipy % conda list
# packages in environment at /Users/czgdp1807/mambaforge/envs/scipy-dev-py311:
#
# Name                    Version                   Build  Channel
accessible-pygments       0.0.5              pyhd8ed1ab_1    conda-forge
alabaster                 0.7.16             pyhd8ed1ab_0    conda-forge
anyio                     4.8.0              pyhd8ed1ab_0    conda-forge
appnope                   0.1.4              pyhd8ed1ab_1    conda-forge
archspec                  0.2.3              pyhd8ed1ab_0    conda-forge
argon2-cffi               23.1.0             pyhd8ed1ab_1    conda-forge
argon2-cffi-bindings      21.2.0          py311h460d6c5_5    conda-forge
array-api-strict          2.1                pyhd8ed1ab_0    conda-forge
arrow                     1.3.0              pyhd8ed1ab_1    conda-forge
asttokens                 3.0.0              pyhd8ed1ab_1    conda-forge
asv                       0.6.4           py311h3f08180_1    conda-forge
asv_runner                0.2.1              pyhd8ed1ab_0    conda-forge
attrs                     24.3.0             pyh71513ae_0    conda-forge
babel                     2.16.0             pyhd8ed1ab_1    conda-forge
beautifulsoup4            4.12.3             pyha770c72_1    conda-forge
beniget                   0.4.2.post1        pyhd8ed1ab_1    conda-forge
bleach                    6.2.0              pyhd8ed1ab_3    conda-forge
bleach-with-css           6.2.0                hd8ed1ab_3    conda-forge
boltons                   24.0.0             pyhd8ed1ab_1    conda-forge
brotli                    1.1.0                hd74edd7_2    conda-forge
brotli-bin                1.1.0                hd74edd7_2    conda-forge
brotli-python             1.1.0           py311h3f08180_2    conda-forge
bzip2                     1.0.8                h99b78c6_7    conda-forge
c-ares                    1.34.4               h5505292_0    conda-forge
c-compiler                1.9.0                hdf49b6b_0    conda-forge
ca-certificates           2024.12.14           hf0a4a13_0    conda-forge
cached-property           1.5.2                hd8ed1ab_1    conda-forge
cached_property           1.5.2              pyha770c72_1    conda-forge
cctools                   1010.6               h4c9edd9_2    conda-forge
cctools_osx-arm64         1010.6               h908b477_2    conda-forge
certifi                   2024.12.14         pyhd8ed1ab_0    conda-forge
cffi                      1.17.1          py311h3a79f62_0    conda-forge
chardet                   5.2.0           py311h267d04e_2    conda-forge
charset-normalizer        3.4.1              pyhd8ed1ab_0    conda-forge
clang                     18.1.8          default_h675cc0c_5    conda-forge
clang-18                  18.1.8          default_h5c12605_5    conda-forge
clang_impl_osx-arm64      18.1.8              h2ae9ea5_23    conda-forge
clang_osx-arm64           18.1.8              h07b0088_23    conda-forge
clangxx                   18.1.8          default_h675cc0c_5    conda-forge
clangxx_impl_osx-arm64    18.1.8              h555f467_23    conda-forge
clangxx_osx-arm64         18.1.8              h07b0088_23    conda-forge
click                     8.1.8              pyh707e725_0    conda-forge
cloudpickle               3.1.1              pyhd8ed1ab_0    conda-forge
colorama                  0.4.6              pyhd8ed1ab_1    conda-forge
colorlog                  6.9.0              pyh707e725_1    conda-forge
comm                      0.2.2              pyhd8ed1ab_1    conda-forge
compiler-rt               18.1.8               h856b3c1_1    conda-forge
compiler-rt_osx-arm64     18.1.8               h832e737_1    conda-forge
compilers                 1.9.0                hce30654_0    conda-forge
conda                     24.11.3         py311h267d04e_0    conda-forge
conda-build               24.11.2         py311h267d04e_1    conda-forge
conda-index               0.5.0              pyhd8ed1ab_0    conda-forge
conda-libmamba-solver     24.11.1            pyhd8ed1ab_0    conda-forge
conda-package-handling    2.4.0              pyh7900ff3_2    conda-forge
conda-package-streaming   0.11.0             pyhd8ed1ab_0    conda-forge
contourpy                 1.3.1           py311h210dab8_0    conda-forge
coverage                  7.6.10          py311h4921393_0    conda-forge
cpp-expected              1.1.0                hffc8910_0    conda-forge
cxx-compiler              1.9.0                hba80287_0    conda-forge
cycler                    0.12.1             pyhd8ed1ab_1    conda-forge
cython                    3.0.11          py311hf7f79b8_3    conda-forge
cython-lint               0.16.6             pyhff2d567_0    conda-forge
debugpy                   1.8.11          py311h155a34a_0    conda-forge
decorator                 5.1.1              pyhd8ed1ab_1    conda-forge
defusedxml                0.7.1              pyhd8ed1ab_0    conda-forge
distlib                   0.3.9              pyhd8ed1ab_1    conda-forge
distro                    1.9.0              pyhd8ed1ab_1    conda-forge
docutils                  0.21.2             pyhd8ed1ab_1    conda-forge
doit                      0.36.0             pyhd8ed1ab_1    conda-forge
entrypoints               0.4                pyhd8ed1ab_1    conda-forge
exceptiongroup            1.2.2              pyhd8ed1ab_1    conda-forge
execnet                   2.1.1              pyhd8ed1ab_1    conda-forge
executing                 2.1.0              pyhd8ed1ab_1    conda-forge
filelock                  3.16.1             pyhd8ed1ab_1    conda-forge
fmt                       11.0.2               h420ef59_0    conda-forge
fonttools                 4.55.3          py311h4921393_1    conda-forge
fortran-compiler          1.9.0                h5692697_0    conda-forge
fqdn                      1.5.1              pyhd8ed1ab_1    conda-forge
freetype                  2.12.1               hadb7bae_2    conda-forge
frozendict                2.4.6           py311hae2e1ce_0    conda-forge
gast                      0.5.5              pyhd8ed1ab_0    conda-forge
gfortran                  13.2.0               h1ca8e4b_1    conda-forge
gfortran_impl_osx-arm64   13.2.0               h252ada1_3    conda-forge
gfortran_osx-arm64        13.2.0               h57527a5_1    conda-forge
gmp                       6.3.0                h7bae524_2    conda-forge
gmpy2                     2.1.5           py311hb5d9ff4_3    conda-forge
greenlet                  3.1.1           py311h155a34a_1    conda-forge
h2                        4.1.0              pyhd8ed1ab_1    conda-forge
hpack                     4.0.0              pyhd8ed1ab_1    conda-forge
hyperframe                6.0.1              pyhd8ed1ab_1    conda-forge
hypothesis                6.124.0            pyha770c72_0    conda-forge
icu                       75.1                 hfee45f7_0    conda-forge
idna                      3.10               pyhd8ed1ab_1    conda-forge
imagesize                 1.4.1              pyhd8ed1ab_0    conda-forge
importlib-metadata        8.5.0              pyha770c72_1    conda-forge
importlib_resources       6.5.2              pyhd8ed1ab_0    conda-forge
iniconfig                 2.0.0              pyhd8ed1ab_1    conda-forge
intersphinx-registry      0.2412.7           pyhd8ed1ab_0    conda-forge
ipykernel                 6.29.5             pyh57ce528_0    conda-forge
ipython                   8.31.0             pyh707e725_0    conda-forge
isl                       0.26            imath32_h347afa1_101    conda-forge
isoduration               20.11.0            pyhd8ed1ab_1    conda-forge
jedi                      0.19.2             pyhd8ed1ab_1    conda-forge
jinja2                    3.1.5              pyhd8ed1ab_0    conda-forge
json5                     0.10.0             pyhd8ed1ab_1    conda-forge
jsonpatch                 1.33               pyhd8ed1ab_1    conda-forge
jsonpointer               3.0.0           py311h267d04e_1    conda-forge
jsonschema                4.23.0             pyhd8ed1ab_1    conda-forge
jsonschema-specifications 2024.10.1          pyhd8ed1ab_1    conda-forge
jsonschema-with-format-nongpl 4.23.0               hd8ed1ab_1    conda-forge
jupyter-cache             1.0.1              pyhff2d567_0    conda-forge
jupyter_client            8.6.3              pyhd8ed1ab_1    conda-forge
jupyter_core              5.7.2              pyh31011fe_1    conda-forge
jupyter_events            0.11.0             pyhd8ed1ab_0    conda-forge
jupyter_server            2.15.0             pyhd8ed1ab_0    conda-forge
jupyter_server_terminals  0.5.3              pyhd8ed1ab_1    conda-forge
jupyterlab_pygments       0.3.0              pyhd8ed1ab_2    conda-forge
jupyterlab_server         2.27.3             pyhd8ed1ab_1    conda-forge
jupyterlite-core          0.5.0              pyh885dcc9_0    conda-forge
jupyterlite-pyodide-kernel 0.5.1              pyh885dcc9_0    conda-forge
jupyterlite-sphinx        0.18.0             pyhd8ed1ab_1    conda-forge
jupytext                  1.16.6             pyh80e38bb_0    conda-forge
kiwisolver                1.4.7           py311h2c37856_0    conda-forge
krb5                      1.21.3               h237132a_0    conda-forge
lcms2                     2.16                 ha0e7c42_0    conda-forge
ld64                      951.9                h4c6efb1_2    conda-forge
ld64_osx-arm64            951.9                hfc0fa09_2    conda-forge
lerc                      4.0.0                h9a09cb3_0    conda-forge
libarchive                3.7.7                h3b16cec_3    conda-forge
libblas                   3.9.0           26_osxarm64_openblas    conda-forge
libbrotlicommon           1.1.0                hd74edd7_2    conda-forge
libbrotlidec              1.1.0                hd74edd7_2    conda-forge
libbrotlienc              1.1.0                hd74edd7_2    conda-forge
libcblas                  3.9.0           26_osxarm64_openblas    conda-forge
libclang-cpp18.1          18.1.8          default_h5c12605_5    conda-forge
libcurl                   8.11.1               h73640d1_0    conda-forge
libcxx                    19.1.7               ha82da77_0    conda-forge
libcxx-devel              18.1.8               h6dc3340_7    conda-forge
libdeflate                1.23                 hec38601_0    conda-forge
libedit                   3.1.20240808    pl5321hafb1f1b_0    conda-forge
libev                     4.33                 h93a5062_2    conda-forge
libexpat                  2.6.4                h286801f_0    conda-forge
libffi                    3.4.2                h3422bc3_5    conda-forge
libgfortran               5.0.0           13_2_0_hd922786_3    conda-forge
libgfortran-devel_osx-arm64 13.2.0               h5d7a38c_3    conda-forge
libgfortran5              13.2.0               hf226fd6_3    conda-forge
libglib                   2.82.2               h07bd6cf_0    conda-forge
libiconv                  1.17                 h0d3ecfb_2    conda-forge
libintl                   0.22.5               h8414b35_3    conda-forge
libjpeg-turbo             3.0.0                hb547adb_1    conda-forge
liblapack                 3.9.0           26_osxarm64_openblas    conda-forge
liblief                   0.14.1               hf9b8971_2    conda-forge
libllvm18                 18.1.8               h5090b49_2    conda-forge
liblzma                   5.6.3                h39f12f2_1    conda-forge
libmamba                  2.0.5                hdf44a08_1    conda-forge
libmambapy                2.0.5           py311hf36a740_1    conda-forge
libnghttp2                1.64.0               h6d7220d_0    conda-forge
libopenblas               0.3.28          openmp_hf332438_1    conda-forge
libpng                    1.6.45               h3783ad8_0    conda-forge
libsodium                 1.0.20               h99b78c6_0    conda-forge
libsolv                   0.7.30               h6c9b7f8_0    conda-forge
libsqlite                 3.47.2               h3f77e49_0    conda-forge
libssh2                   1.11.1               h9cc3647_0    conda-forge
libtiff                   4.7.0                h551f018_3    conda-forge
libwebp-base              1.5.0                h2471fea_0    conda-forge
libxcb                    1.17.0               hdb1d25a_0    conda-forge
libxml2                   2.13.5               h178c5d8_1    conda-forge
libzlib                   1.3.1                h8359307_2    conda-forge
llvm-openmp               19.1.7               hdb05f8b_0    conda-forge
llvm-tools                18.1.8               h5090b49_2    conda-forge
llvm-tools-18             18.1.8               h5090b49_2    conda-forge
lz4-c                     1.10.0               h286801f_1    conda-forge
lzo                       2.10              h93a5062_1001    conda-forge
markdown-it-py            3.0.0              pyhd8ed1ab_1    conda-forge
markupsafe                3.0.2           py311h4921393_1    conda-forge
matplotlib                3.10.0          py311ha1ab1f8_0    conda-forge
matplotlib-base           3.10.0          py311h031da69_0    conda-forge
matplotlib-inline         0.1.7              pyhd8ed1ab_1    conda-forge
mdit-py-plugins           0.4.2              pyhd8ed1ab_1    conda-forge
mdurl                     0.1.2              pyhd8ed1ab_1    conda-forge
menuinst                  2.2.0           py311h267d04e_0    conda-forge
meson                     1.6.1              pyhd8ed1ab_0    conda-forge
meson-python              0.17.1             pyh70fd9c4_1    conda-forge
mistune                   3.1.0              pyhd8ed1ab_0    conda-forge
more-itertools            10.6.0             pyhd8ed1ab_0    conda-forge
mpc                       1.3.1                h8f1351a_1    conda-forge
mpfr                      4.2.1                hb693164_3    conda-forge
mpmath                    1.3.0              pyhd8ed1ab_1    conda-forge
munkres                   1.1.4              pyh9f0ad1d_0    conda-forge
mypy                      1.14.1          py311h917b07b_0    conda-forge
mypy_extensions           1.0.0              pyha770c72_1    conda-forge
myst-nb                   1.1.2              pyhd8ed1ab_1    conda-forge
myst-parser               4.0.0              pyhd8ed1ab_1    conda-forge
nbclient                  0.10.2             pyhd8ed1ab_0    conda-forge
nbconvert-core            7.16.5             pyhd8ed1ab_1    conda-forge
nbformat                  5.10.4             pyhd8ed1ab_1    conda-forge
ncurses                   6.5                  h5e97a16_2    conda-forge
nest-asyncio              1.6.0              pyhd8ed1ab_1    conda-forge
ninja                     1.12.1               h420ef59_0    conda-forge
nlohmann_json             3.11.3               h00cdb27_1    conda-forge
numpy                     2.2.1           py311h4b914e2_0    conda-forge
numpydoc                  1.8.0              pyhd8ed1ab_1    conda-forge
openblas                  0.3.28          openmp_hea878ba_1    conda-forge
openjpeg                  2.5.3                h8a3d83b_0    conda-forge
openssl                   3.4.0                h81ee809_1    conda-forge
overrides                 7.7.0              pyhd8ed1ab_1    conda-forge
packaging                 24.2               pyhd8ed1ab_2    conda-forge
pandocfilters             1.5.0              pyhd8ed1ab_0    conda-forge
parso                     0.8.4              pyhd8ed1ab_1    conda-forge
patch                     2.7.6             h27ca646_1002    conda-forge
pcre2                     10.44                h297a79d_2    conda-forge
pexpect                   4.9.0              pyhd8ed1ab_1    conda-forge
pickleshare               0.7.5           pyhd8ed1ab_1004    conda-forge
pillow                    11.1.0          py311hb9ba9e9_0    conda-forge
pip                       24.3.1             pyh8b19718_2    conda-forge
pkg-config                0.29.2            hde07d2e_1009    conda-forge
pkginfo                   1.12.0             pyhd8ed1ab_1    conda-forge
pkgutil-resolve-name      1.3.10             pyhd8ed1ab_2    conda-forge
platformdirs              4.3.6              pyhd8ed1ab_1    conda-forge
pluggy                    1.5.0              pyhd8ed1ab_1    conda-forge
ply                       3.11               pyhd8ed1ab_3    conda-forge
pooch                     1.8.2              pyhd8ed1ab_1    conda-forge
prometheus_client         0.21.1             pyhd8ed1ab_0    conda-forge
prompt-toolkit            3.0.48             pyha770c72_1    conda-forge
psutil                    6.1.1           py311h917b07b_0    conda-forge
pthread-stubs             0.4               hd74edd7_1002    conda-forge
ptyprocess                0.7.0              pyhd8ed1ab_1    conda-forge
pure_eval                 0.2.3              pyhd8ed1ab_1    conda-forge
py-lief                   0.14.1          py311h3f08180_2    conda-forge
pybind11                  2.13.6             pyh1ec8472_2    conda-forge
pybind11-abi              4                    hd8ed1ab_3    conda-forge
pybind11-global           2.13.6             pyh415d2e4_2    conda-forge
pycodestyle               2.12.1             pyhd8ed1ab_1    conda-forge
pycosat                   0.6.6           py311h917b07b_2    conda-forge
pycparser                 2.22               pyh29332c3_1    conda-forge
pydata-sphinx-theme       0.16.1             pyhd8ed1ab_0    conda-forge
pydevtool                 0.3.0              pyhd8ed1ab_0    conda-forge
pygments                  2.19.1             pyhd8ed1ab_0    conda-forge
pympler                   1.1                pyhd8ed1ab_1    conda-forge
pyobjc-core               11.0            py311hab620ed_0    conda-forge
pyobjc-framework-cocoa    11.0            py311hab620ed_0    conda-forge
pyparsing                 3.2.1              pyhd8ed1ab_0    conda-forge
pyproject-metadata        0.9.0              pyhd8ed1ab_1    conda-forge
pysocks                   1.7.1              pyha55dd90_7    conda-forge
pytest                    8.3.4              pyhd8ed1ab_1    conda-forge
pytest-cov                6.0.0              pyhd8ed1ab_1    conda-forge
pytest-timeout            2.3.1              pyhd8ed1ab_2    conda-forge
pytest-xdist              3.6.1              pyhd8ed1ab_1    conda-forge
python                    3.11.11         hc22306f_1_cpython    conda-forge
python-dateutil           2.9.0.post0        pyhff2d567_1    conda-forge
python-fastjsonschema     2.21.1             pyhd8ed1ab_0    conda-forge
python-json-logger        2.0.7              pyhd8ed1ab_0    conda-forge
python-libarchive-c       5.1             py311h267d04e_1    conda-forge
python_abi                3.11                    5_cp311    conda-forge
pythran                   0.17.0             pyh27800b7_1    conda-forge
pytz                      2024.2             pyhd8ed1ab_1    conda-forge
pyyaml                    6.0.2           py311h460d6c5_1    conda-forge
pyzmq                     26.2.0          py311h730b646_3    conda-forge
qhull                     2020.2               h420ef59_5    conda-forge
readline                  8.2                  h92ec313_1    conda-forge
referencing               0.35.1             pyhd8ed1ab_1    conda-forge
reproc                    14.2.5.post0         h5505292_0    conda-forge
reproc-cpp                14.2.5.post0         h286801f_0    conda-forge
requests                  2.32.3             pyhd8ed1ab_1    conda-forge
rfc3339-validator         0.1.4              pyhd8ed1ab_1    conda-forge
rfc3986-validator         0.1.1              pyh9f0ad1d_0    conda-forge
rich                      13.9.4             pyhd8ed1ab_1    conda-forge
rich-click                1.8.5              pyhd8ed1ab_0    conda-forge
ripgrep                   14.1.1               h0ef69ab_0    conda-forge
rpds-py                   0.22.3          py311h3ff9189_0    conda-forge
ruamel.yaml               0.18.10         py311h917b07b_0    conda-forge
ruamel.yaml.clib          0.2.8           py311hae2e1ce_1    conda-forge
ruff                      0.9.1           py311hdb0c05a_0    conda-forge
send2trash                1.8.3              pyh31c8845_1    conda-forge
setuptools                67.1.0             pyhd8ed1ab_0    conda-forge
sigtool                   0.1.3                h44b9a77_0    conda-forge
simdjson                  3.11.6               ha393de7_0    conda-forge
six                       1.17.0             pyhd8ed1ab_0    conda-forge
sniffio                   1.3.1              pyhd8ed1ab_1    conda-forge
snowballstemmer           2.2.0              pyhd8ed1ab_0    conda-forge
sortedcontainers          2.4.0              pyhd8ed1ab_0    conda-forge
soupsieve                 2.5                pyhd8ed1ab_1    conda-forge
spdlog                    1.15.0               h096ffd4_0    conda-forge
sphinx                    7.4.7              pyhd8ed1ab_0    conda-forge
sphinx-copybutton         0.5.2              pyhd8ed1ab_1    conda-forge
sphinx-design             0.6.1              pyhd8ed1ab_2    conda-forge
sphinxcontrib-applehelp   2.0.0              pyhd8ed1ab_1    conda-forge
sphinxcontrib-devhelp     2.0.0              pyhd8ed1ab_1    conda-forge
sphinxcontrib-htmlhelp    2.1.0              pyhd8ed1ab_1    conda-forge
sphinxcontrib-jsmath      1.0.1              pyhd8ed1ab_1    conda-forge
sphinxcontrib-qthelp      2.0.0              pyhd8ed1ab_1    conda-forge
sphinxcontrib-serializinghtml 1.1.10             pyhd8ed1ab_1    conda-forge
sqlalchemy                2.0.37          py311h917b07b_0    conda-forge
stack_data                0.6.3              pyhd8ed1ab_1    conda-forge
tabulate                  0.9.0              pyhd8ed1ab_2    conda-forge
tapi                      1300.6.5             h03f4b80_0    conda-forge
terminado                 0.18.1             pyh31c8845_0    conda-forge
threadpoolctl             3.5.0              pyhc1e730c_0    conda-forge
tinycss2                  1.4.0              pyhd8ed1ab_0    conda-forge
tk                        8.6.13               h5083fa2_1    conda-forge
tokenize-rt               6.1.0              pyhd8ed1ab_1    conda-forge
toml                      0.10.2             pyhd8ed1ab_1    conda-forge
tomli                     2.2.1              pyhd8ed1ab_1    conda-forge
tornado                   6.4.2           py311h917b07b_0    conda-forge
tqdm                      4.67.1             pyhd8ed1ab_1    conda-forge
traitlets                 5.14.3             pyhd8ed1ab_1    conda-forge
truststore                0.10.0             pyhd8ed1ab_0    conda-forge
types-psutil              6.1.0.20241221     pyhd8ed1ab_0    conda-forge
types-python-dateutil     2.9.0.20241206     pyhd8ed1ab_0    conda-forge
typing-extensions         4.12.2               hd8ed1ab_1    conda-forge
typing_extensions         4.12.2             pyha770c72_1    conda-forge
typing_utils              0.1.0              pyhd8ed1ab_1    conda-forge
tzdata                    2024b                hc8b5060_0    conda-forge
unicodedata2              16.0.0          py311h917b07b_0    conda-forge
uri-template              1.3.0              pyhd8ed1ab_1    conda-forge
urllib3                   2.3.0              pyhd8ed1ab_0    conda-forge
virtualenv                20.29.0            pyhd8ed1ab_0    conda-forge
wcwidth                   0.2.13             pyhd8ed1ab_1    conda-forge
webcolors                 24.11.1            pyhd8ed1ab_0    conda-forge
webencodings              0.5.1              pyhd8ed1ab_3    conda-forge
websocket-client          1.8.0              pyhd8ed1ab_1    conda-forge
wheel                     0.45.1             pyhd8ed1ab_1    conda-forge
xorg-libxau               1.0.12               h5505292_0    conda-forge
xorg-libxdmcp             1.1.5                hd74edd7_0    conda-forge
yaml                      0.2.5                h3422bc3_2    conda-forge
yaml-cpp                  0.8.0                h13dd4ca_0    conda-forge
zeromq                    4.3.5                hc1bb282_7    conda-forge
zipp                      3.21.0             pyhd8ed1ab_1    conda-forge
zlib                      1.3.1                h8359307_2    conda-forge
zstandard                 0.23.0          py311ha60cc69_1    conda-forge
zstd                      1.5.6                hb46c0d2_0    conda-forge
```

</details>

Machine details - Apple Macbook Pro (Apple M2 Pro)

j-bowhay commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -112,6 +112,9 @@ def laplacian(
     symmetric if the input csgraph was symmetric.
 
+    The algorithmic complexity is for a graph with V
+    vertices is ``O(V^2)``.
```
Ah sorry I meant we should make all of them monospace not just this one

czgdp1807 replied:
Okay. I thought only this specific location is causing issues. Sure. Give me a few minutes, I will update everywhere and push.

czgdp1807 replied:
Addressed.

czgdp1807 committed:
DOC: Address reviews

czgdp1807 committed:
DOC: Address reviews

czgdp1807 commented:
Okay. The tests pass. :-)

czgdp1807 commented:
> Are you getting these complexities based on published papers?

All the time complexities are taken from Wikipedia. @dschult

j-bowhay commented on scipy/sparse/csgraph/_shortest_path.pyx:
```diff
@@ -876,6 +883,10 @@ def bellman_ford(csgraph, directed=True, indices=None,
     If multiple valid solutions are possible, output may vary with SciPy and
     Python version.
 
+    Computational cost is approximately ``O(V * E)``, where
```
Although this matches Wikipedia, it would be worth checking the blame to try workout the reason for this discrepancy

czgdp1807 replied:
`k` is defined as average number of connected edges per node, which is average degree. Actually, this algorithm is O(V * E). We also define E as O(V^2). So, O(V^3). The documentation describes, O(V^3 * k). I think O(V * E) is more accurate because it doesn't assume E as O(V^2). In the worst case number of edges i.e., E is O(V^2). In a sparsely connected E can also be O(V). E can be a any function of number of nodes i.e, V. In the worst case it is V^2 but not always. 


This raises a question that how is Bellman Ford algorithm implemented in SciPy. I will look into the blame as well as the code. There must be a good reason to go for O(N^3 k) time complexity (may be the implementation is like that).

czgdp1807 replied:
Well, yes the implementation is like that,

```zsh
for i in range(Nind):
        j_source = source_indices[i]

        # relax all edges N-1 times
        for count in range(N - 1):
            for j in range(N):
                d1 = dist_matrix[i, j]
                for k in range(csr_indptr[j], csr_indptr[j + 1]):
                    w12 = csr_weights[k]
                    ind_k = csr_indices[k]
                    d2 = dist_matrix[i, ind_k]
                    if d1 + w12 < d2:
                        dist_matrix[i, ind_k] = d2 = d1 + w12
                        if return_pred:
                            pred[i, ind_k] = j
                    if d2 + w12 < d1:
                        dist_matrix[i, j] = d1 = d2 + w12
                        if return_pred:
                            pred[i, j] = ind_k
```

This translates to Nind * (N - 1) * N * k, which is O(N^3 k), only when Nind = N. If Nind is not equal to N then, the correct time complexity would be O(I * N^2 * k), where I = len(indices).

Seems like we will have to update the time complexity for this algorithm at both the places (the change I made and also in the link you sent). Specifically mention that when I = N, then time complexity would be O(N * N^2 * k).

czgdp1807 replied:
@j-bowhay Let me know if you think the same. I will make the update then.

j-bowhay replied:
Seems reasonable thanks

czgdp1807 replied:
Done. I have also added backticks wherever needed in my changes.

rgommers commented:
> * [csgraph_from_dense](https://docs.scipy.org/doc/scipy-1.15.0/reference/generated/scipy.sparse.csgraph.csgraph_from_dense.html#scipy.sparse.csgraph.csgraph_from_dense) - Do we need to mention time complexity for this?

I agree that that isn't needed. The goal of the issue was to get complexities for graph algorithms, not for utility functions.

czgdp1807 committed:
DOC: Fix time complexity of bellman ford algorithm

czgdp1807 committed:
DOC: Added backticks at all the necessary places

czgdp1807 committed:
[docs only]

czgdp1807 commented:
@j-bowhay If you find my latest commits then please let me know. I will drop the last commit in that case and run the full CI.

windows-server-2003 commented:
I would love to see this PR merged. Thank you @czgdp1807!  
Although I am not a maintainer, I reviewed the changes. Hopefully, this helps the merging process of this PR.  
I would appreciate if @czgdp1807 takes a look at the suggestions.

I think the complexity in the documentations should match the implementation rather than the academic results or well-known complexity of the algorithm, so I reviewed from that point of view.

Below are some suggestions:

### 1. Shortest path related
1. 🟡 Dijkstra  
  Although the time bound is correct, when `min_only` is true, the complexity is `O((E + V) * log(V))` no matter how large `indices` is.  
  Also, I think the word `approximately` should be removed, because it is a solid, provable time bound.
2. 🟠 Johnson  
  Johnson is essentially 1x Bellman-Ford + |indices|x Dijkstra, so I think `O(I * (E + V) * log(V) + V * E)` would be a more accurate bound (note the `E`  factor in the first term).
3. 🟡 Bellman-Ford  
  Although the time bound is correct, I think `O(I * V * E)` is a simpler and consistent expression (eliminating `k` entirely).  
  I also think the `If ``I = N``...` statement is unnecessary.

### 2. Typos / Notation consistency
 - 🟡 `floyd_warshall`, `connected_components`: There are double `is`
 - 🔵 `minimum_spanning_tree`: ```...a graph with ``V`` vertices and ``E`` edges is...``` would be more concise
 - 🔵 `minimum_spanning_tree`, `yen`: Spacing before/after `*` would make it more consistent
 - 🔵 `shortest_path`, `dijkstra`, `bellman_ford`: Using `V` instead of `N` would make it more consistent
 - 🔵 `shortest_path`, `bellman_ford`: Using `O()` instead of `O[]` would make it more consistent

### 3. 🔵 Dense input related
This is more like a nitpick, but most functions cost an additional $O(V^2)$ when the input is a dense array-like because it is first converted to CSR (sparse expression).  
Note that an array-like input can represent a sparse graph (though not efficiently), which means we cannot ignore $O(V^2)$ just because $O(E)$ (or a larger term) is in the main complexity.

It might be better to add a short notice like
```
*array_like* ``csgraph`` input incurs an additional O(V^2) cost.
```

Below are the functions that require an additional $O(V^2)$ when the input is array-like:

 - `laplacian` (also when form='array')
 - `minimum_spanning_tree`
 - `shortest_path`, `dijkstra`
 - `connected_components`
 - `breadth_first_order`, `breadth_first_tree`
 - `depth_first_order`, `depth_first_tree`
 - `reconstruct_path`

Below are the functions that the main complexity dominates the $O(V^2)$ term:
 - `floyd_warshall`, `bellman_ford`, `johnson`, `yen`
 - `construct_dist_matrix`

Below are the functions that cannot take array-like input in the first place:
 - `maximum_bipartite_matching`, `min_weight_full_bipartite_matching`, (+`maximum_flow`)
 - `reverse_cuthill_mckee`
 - `structural_rank`
